src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -1225,12 +1226,12 @@ void MacroAssembler::lookup_interface_method_stub(Register recv_klass,
 
   ldrw(scan_temp, Address(recv_klass, Klass::vtable_length_offset()));
   add(recv_klass, recv_klass, vtable_start_offset + ioffset);
-  // itableOffsetEntry[] itable = recv_klass + Klass::vtable_start_offset() + recv_klass->_vtable_len;
+  // itableOffsetEntry[] itable = recv_klass + Klass::vtable_start_offset() + sizeof(vtableEntry) * recv_klass->_vtable_len;
   // temp_itbl_klass = itable[0]._interface;
-  ldr(temp_itbl_klass, Address(recv_klass, scan_temp, Address::lsl(3)));
+  ldr(temp_itbl_klass, Address(recv_klass, scan_temp, Address::lsl(exact_log2(vtableEntry::size_in_bytes()))));
```

Comment:
The shift in `ldr` must be 3.
Can we have a static assert checking that? There is an assert in `Address::encode` which will catch the issue in debug builds.

Reply:
I'm very surprised that isn't a guarantee. I had to look to be sure. I'd simply change `assert(ext().shift() <= 0 || ext().shift() == (int)size, "bad shift");` to a `guarantee`.

Reply:
I want to avoid modifying the assembler_aarch64.hpp file within this change, so I add guarantee in lookup_interface_method.
By the way, [assembler_aarch64.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/assembler_aarch64.hpp) contains 100+ assert calls and 34 guarantee calls. I think more cases can be changed, although I am not sure what the rule is for using this or that call.

Reply:
> I want to avoid modifying the assembler_aarch64.hpp file within this change, so I add guarantee in lookup_interface_method. By the way, [assembler_aarch64.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/assembler_aarch64.hpp) contains 100+ assert calls and 34 guarantee calls. I think more cases can be changed, although I am not sure what the rule is for using this or that call.

A JBS issue?

Reply:
No, I don't think so.

The balance between `assert()` and `guarantee()` is a matter of judgement, not hard rules. 

We don't want to do consistency checking at runtime if we can avoid it, because it adds to startup time.

The `guarantee()` that an operand fits into its field in an instruction has proved its worth, many times. This is because AArch64 has a variety of offset fields, and these are  of varying lengths. It's very easy to be caught out, and sometimes this has happened even after release.

In most other places, an `assert()` is all we need. This is because the thing being asserted doesn't vary much, and we only need to check it once, when we run a debug build.

In the case of checking that the shift in a memory-access instruction matches the operand size, the check will be inlined. I haven't looked, but I expect it will be eliminated by the C++ optimizer at compile time.

In the case of the ldr instruction here, the operand size is determined already by the `ldr` instruction: it's `wordSize`. It cannot be anything else. The AArch64 back end asserts that elsewhere: 

```
  int vte_size    = vtableEntry::size_in_bytes();
  assert(vte_size == wordSize, "else adjust times_vte_scale");
```

Reply:
I agree. I keep assert in MacroAssembler::lookup_interface_method_stub and change assert to guarantee in Address::encode.